src/dinit-iostream.cc

```diff
@@ -0,0 +1,748 @@
+#include <limits>
+#include <cstdio>
+#include <cstring>
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <dinit-iostream.h>
+#include <baseproc-sys.h>
+
+// See dinit-iostream.h for inforamtion about the usage
+
+namespace {
+
+// A simple trick for calculating how much digits can be in a type with compile-time integer
+// logarithm like function.
+template <typename T> constexpr unsigned type_max_numdigits(T num = std::numeric_limits<T>::max(),
+        unsigned pow = 0)
+{
+    return (num == 0) ? pow : type_max_numdigits(num / 10, pow + 1);
+}
+
+// We should have space for sign in signed variables.
+constexpr unsigned UINT_MAX_CHARS = type_max_numdigits<unsigned>();
+constexpr unsigned INT_MAX_CHARS = type_max_numdigits<int>() + 1;
+constexpr unsigned ULONG_MAX_CHARS = type_max_numdigits<unsigned long>();
+constexpr unsigned LONG_MAX_CHARS = type_max_numdigits<long>() + 1;
+
+} /* anonymous namespace */
+
+namespace dio {
+
+// class io_base
+
+streambuf *io_base::get_buf() noexcept
+{
+    return buf.get();
+}
+
+bool io_base::is_open() noexcept
+{
+    return (fd >= 0);
+}
+
+void io_base::set_fd(const int newfd) noexcept
+{
+    fd = newfd;
+}
+
+int io_base::get_fd() noexcept
+{
+    return fd;
+}
+
+// class ostream
+
+ssize_t ostream::put(const char *msg, size_t count) noexcept
+{
+    if (!good()) {
+        return -1;
+    }
+
+    if (count == 0) {
+        return 0;
+    }
+    if (count > buf->get_free()) {
+        struct iovec bufv[3];
+        char *ptr = buf->get_ptr(0);
+
+        bufv[0].iov_base = ptr;
+        bufv[0].iov_len = (buf->get_length() > 0) ? buf->get_contiguous_length(ptr) : 0;
+        if (buf->get_length() != bufv[0].iov_len) {
+            bufv[1].iov_base = buf->get_buf_base();
+            bufv[1].iov_len = buf->get_length() - bufv[0].iov_len;
+        }
+        else {
+            bufv[1].iov_base = nullptr;
+            bufv[1].iov_len = 0;
+        }
+        bufv[2].iov_base = const_cast<char *>(msg);
+        bufv[2].iov_len = count;
+
+        unsigned buffer_length = bufv[0].iov_len + bufv[1].iov_len;
+
+        ssize_t ret = bp_sys::writev(fd, bufv, 3);
+        if (ret < 0) {
+            io_error = errno;
+            return -1;
+        }
+        if (ret >= buffer_length) {
+            buf->consume(buffer_length);
+            ret -= buffer_length;
+        }
+        else {
+            buf->consume(ret);
+            ret = 0;
+        }
+        if (ret != count) {
+            io_error = errno;
```

Comment:
Partial write is not an error, as I've said before. `errno` won't be set.

This needs to buffer the remaining part of the message if possible, or call `write` (or `writev`) again if not (to write the remaining part).